pennylane/qchem/vibrational/bosonic.py

```diff
@@ -0,0 +1,624 @@
+        r"""Multiply a BoseWord with another BoseWord, a BoseSentence, or a constant.
+
+        >>> w = BoseWord({(0, 0) : '+', (1, 1) : '-'})
+        >>> print(w * w)
+        b⁺(0) b(1) b⁺(0) b(1)
+        """
+
+        if isinstance(other, BoseWord):
+            if len(self) == 0:
+                return copy(other)
+
+            if len(other) == 0:
+                return copy(self)
+
+            order_final = [i[0] + len(self) for i in other.sorted_dic.keys()]
+            other_wires = [i[1] for i in other.sorted_dic.keys()]
+
+            dict_other = dict(
+                zip(
+                    [(order_idx, other_wires[i]) for i, order_idx in enumerate(order_final)],
+                    other.values(),
+                )
+            )
+            dict_self = dict(zip(self.keys(), self.values()))
+
+            dict_self.update(dict_other)
+
+            return BoseWord(dict_self)
+
+        if isinstance(other, BoseSentence):
+            return BoseSentence({self: 1}) * other
+
+        if not isinstance(other, TensorLike):
+            raise TypeError(f"Cannot multiply BoseWord by {type(other)}.")
+
+        if qml.math.size(other) > 1:
+            raise ValueError(
+                f"Arithmetic Bose operations can only accept an array of length 1, "
+                f"but received {other} of length {len(other)}"
+            )
+
+        return BoseSentence({self: other})
+
+    def __rmul__(self, other):
+        r"""Reverse multiply a BoseWord
+
+        Multiplies a BoseWord "from the left" with an object that can't be modified
+        to support __mul__ for BoseWord. Will be defaulted in for example
+        ``2 * BoseWord({(0, 0): "+"})``, where the ``__mul__`` operator on an integer
+        will fail to multiply with a BoseWord"""
+
+        return self.__mul__(other)
+
+    def __pow__(self, value):
+        r"""Exponentiate a Bose word to an integer power.
+
+        >>> w = BoseWord({(0, 0) : '+', (1, 1) : '-'})
+        >>> print(w**3)
+        b⁺(0) b(1) b⁺(0) b(1) b⁺(0) b(1)
+        """
+
+        if value < 0 or not isinstance(value, int):
+            raise ValueError("The exponent must be a positive integer.")
+
+        operator = BoseWord({})
+
+        for _ in range(value):
+            operator *= self
+
+        return operator
+
+    def normal_order(self):
+        r"""Convert a BoseWord to its normal-ordered form.
+
+        >>> bw = BoseWord({(0, 0): "-", (1, 0): "-", (2, 0): "+", (3, 0): "+"})
+        >>> print(bw.normal_order())
+        4.0 * b⁺(0) b(0)
+        + 2.0 * I
+        + 1.0 * b⁺(0) b⁺(0) b(0) b(0)
+        """
+        bw_terms = sorted(self)
+        len_op = len(bw_terms)
+        bw_comm = BoseSentence({BoseWord({}): 0.0}, is_hardcore=self.is_hardcore)
+
+        if len_op == 0:
+            return 1 * BoseWord({})
+
+        bw = self
+
+        l = 0
+        for r in range(len_op):
+            if self[bw_terms[r]] == "+":
+                if l == r:
```

Comment:
What does this branch do?

Reply:
For the two pointer solution, we have `l` and `r` starting at `idx 0`. Suppose we have `b1^ b1^ b0 b0 b1^ b1^`, we want to increment `l` until we reach `b0`. Two solns exist, one is basically to write a while loop outside of this loop that says, `while self[bw[terms[r]] == "+" l++ r++`, the other is to write this branch.